src/dot_parser/mod.rs

```diff
@@ -0,0 +1,299 @@
+                        let merged = PyDict::new(py);
+                        for (k, v) in default_node_attrs.iter() {
+                            merged.set_item(k, v)?;
+                        }
+                        for maybe_attr in it {
+                            if maybe_attr.as_rule() == Rule::attr_list {
+                                let map = parse_attr_list_to_map(maybe_attr);
+                                for (k, v) in map {
+                                    merged.set_item(k, v)?;
+                                }
+                            }
+                        }
+                        node_attrs_map.insert(name.clone(), merged.into());
+                    }
+
+                    Rule::edge_stmt => {
+                        // edge_stmt
+                        let mut endpoints: Vec<String> = Vec::new();
+                        let mut operators: Vec<String> = Vec::new();
+
+                        // start collected edge attrs from defaults
+                        let collected = PyDict::new(py);
+                        for (k, v) in default_edge_attrs.iter() {
+                            collected.set_item(k, v)?;
+                        }
+
+                        for child in stmt.into_inner() {
+                            match child.as_rule() {
+                                Rule::edge_point => {
+                                    for ep_child in child.into_inner() {
+                                        if ep_child.as_rule() == Rule::node_id {
+                                            let n = node_id_to_string(ep_child);
+                                            endpoints.push(n);
+                                        }
+                                    }
+                                }
+                                Rule::edge_op => {
+                                    operators.push(child.as_str().trim().to_string());
+                                }
+                                Rule::attr_list => {
+                                    let map = parse_attr_list_to_map(child);
+                                    for (k, v) in map {
+                                        collected.set_item(k, v)?;
+                                    }
+                                }
+                                _ => {}
+                            }
+                        }
+
+                        // create pairwise edges and add them to the graph
+                        for i in 0..endpoints.len().saturating_sub(1) {
+                            let src = endpoints[i].clone();
+                            let dst = endpoints[i + 1].clone();
+
+                            let src_idx = *node_map.entry(src.clone()).or_insert_with(|| {
+                                let py_node = PyString::new(py, &src).into();
+                                if is_directed {
+                                    di_graph.as_mut().unwrap().graph.add_node(py_node)
```

Comment:
If you follow the previous recommendation, this becomes `graph.add_node` which arguably is simpler that handling the `Option<PyGraph>` each time